### PR TITLE
Fix compatibility issues with various versions of Rust.

### PR DESCRIPTION
**Pull request by Anders429**
Fix for #5. This allows the crate to be used in all versions of Rust, from `1.0.0` upwards, as specified by the Rust version badge. I have also added coverage for each of these versions to the CI, so that they don't accidentally break again in the future :)

I had to remove the Kleene operator (`?`) from `assert_matches!` in order to not lose coverage in `rustc 1.32.0`, since the operator requires edition 2018 in order to be used before `rustc 1.37.0`. I simply broke the optional arguments into separate matchers. Interestingly, this allows `assert_matches!` to be used in earlier `rustc` versions. I bumped the MSRV for it back to `1.26.0`, because that's as far back as it tests, but you could probably move it even further back by making the tests compatible (it builds all the way back to version `1.7.0`, and the only thing holding it back is that the tests aren't compatible before `1.26.0`). For now, I've left it at `1.26.0`.

### CODE REVIEW

**Comment by Anders429** [comment]
Hmm, I think the `no_std` failure is a shortcoming of `cargo-nono`. Building the crate on a `no_std` target works just fine:

```
$ cargo build --target thumbv6m-none-eabi
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
```

Seems like `cargo-nono` doesn't check for anything besides `#![no_std]` or something like `#![cfg_attr(not(feature = "std"), no_std)]`. Going off of [this issue](https://github.com/hobofan/cargo-nono/issues/61), it doesn't seem like other cases are covered in its checks.

**Comment by Anders429** [comment]
After further investigation, the breakage is indeed due to a limitation of `cargo-nono`. [The code](https://github.com/hobofan/cargo-nono/blob/master/src/check_source.rs#L222-L230) only checks for `#![no_std]` and `#![cfg_attr(not(test), no_std)]` explicitly, and doesn't check for any other `cfg_attr` form. There is [one other check](https://github.com/hobofan/cargo-nono/blob/d834f1e2d61c0e1d881a1931c526570d02e8c0ca/src/check_source.rs#L178-L189) for features that disable `no_std`, but that is all.

**Comment by Anders429** [comment]
I've changed the no_std test in the CI to directly build the crate on a `no_std` target, `thumbv6m-none-eabi`. This is, in my opinion, better than running `cargo-nono`, as it actually runs the build and will fail if there are any actual errors in building on `no_std`.

This is the same approach used by many other crates, [including `rand`](https://github.com/rust-random/rand/blob/6ecbe2626b2cc6110a25c97b1702b347574febc7/.github/workflows/test.yml#L160-L172).

If you'd rather not do this, that's fine with me, but this resolves the CI failure while still showing `no_std` compatibility :)

File: src/assert_matches.rs [hunk 1]
```diff
@@ -41,23 +41,43 @@
...
             other => {
                 panic!(r#"assertion failed, expression does not match any of the given variants.
     expression: {:?}
-    variants: {}"#, other, stringify!($($pattern) |+ $(if $guard)?));
+    variants: {}"#, other, stringify!($($pattern) |+));
             }
         }
     };
-    ($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?, $($arg:tt)+) => {
+    ($expression:expr, $( $pattern:pat )|+ if $guard: expr) => {
```
Comment (svartalf) [inline src/assert_matches.rs:54]: I'm not that good in macros, but are we ensure that former `$(if $guard)?` clause is fully covered by these match branches?

**Review by svartalf — CHANGES_REQUESTED** [review]
> If you'd rather not do this, that's fine with me

Also fine by me, thanks for doing it!
Can you elaborate on one thing I personally find confusing?

File: src/assert_matches.rs (hunk 1 above)
Comment (Anders429) [reply src/assert_matches.rs:54]: The Kleene operator `?` means at most one repetition ([source](https://doc.rust-lang.org/edition-guide/rust-2018/macros/at-most-once.html)). So it basically can match two cases: either the `if $guard` being present or it being absent. Basically, it's equivalent to writing out two matchers, one to match each case.

A simpler example (from the above source) makes it pretty clear that

```rust
macro_rules! foo {
    ($a:ident, $b:expr) => {
        println!("{}", $a);
        println!("{}", $b);
    };
    ($a:ident) => {
        println!("{}", $a);
    }
}
```

is equivalent to

```rust
macro_rules! foo {
    ($a:ident $(, $b:expr)?) => {
        println!("{}", $a);

        $(
            println!("{}", $b);
         )?
    }
}
```

because the `?` matches the optional presence of `b`. Essentially, what I'm trying to say is that for every Kleene operator used, you could rewrite the same macro with the Kleene operator gone, but would need twice as many matchers.

So in the `assert_matches!` case, the number of matchers doubles. The original matchers were

```rust
($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?) => ..
($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?, $($arg:tt)+) => ..
```

and in the equivalent Kleene-less case we end up with twice as many matchers:

```rust
($expression:expr, $( $pattern:pat )|+) => ..
($expression:expr, $( $pattern:pat )|+ if $guard: expr) => ..
($expression:expr, $( $pattern:pat )|+, $($arg:tt)+) => ..
($expression:expr, $( $pattern:pat )|+ if $guard: expr, $($arg:tt)+) => ..
```

Those two code blocks cover all of the same cases :)

So, tl;dr: yes, I believe we can be sure that all cases are fully covered.

Comment (svartalf) [reply src/assert_matches.rs:54]: Alright, thanks for this huge explanation 👍🏼

**Review by svartalf — APPROVED** [review]

**Comment by svartalf** [comment]
I'll publish new version in a couple hours, thanks for your work!

**Comment by svartalf** [comment]
Follow-up: your changes were released in [`claim = "0.5.0"`](https://crates.io/crates/claim/0.5.0) release
